python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1110 @@
+
+    cdef CScalarExpression* scalar
+
+    def __init__(self, value):
+        cdef:
+            shared_ptr[CScalar] scalar
+            shared_ptr[CScalarExpression] expression
+
+        if isinstance(value, bool):
+            scalar = MakeScalar(<c_bool>value)
+        elif isinstance(value, float):
+            scalar = MakeScalar(<double>value)
+        elif isinstance(value, int):
+            scalar = MakeScalar(<int64_t>value)
+        else:
+            raise TypeError('Not yet supported scalar value: {}'.format(value))
+
+        expression.reset(new CScalarExpression(scalar))
+        self.init(<shared_ptr[CExpression]> expression)
+
+    cdef void init(self, const shared_ptr[CExpression]& sp):
+        Expression.init(self, sp)
+        self.scalar = <CScalarExpression*> sp.get()
+
+    # TODO(kszucs): implement once we have proper Scalar bindings
+    # @property
+    # def value(self):
+    #     return pyarrow_wrap_scalar(self.scalar.value())
+
+
+cdef class FieldExpression(Expression):
+
+    cdef CFieldExpression* scalar
+
+    def __init__(self, name):
+        cdef:
+            c_string field_name = tobytes(name)
+            shared_ptr[CExpression] expression
+        expression.reset(new CFieldExpression(field_name))
+        self.init(expression)
+
+    cdef void init(self, const shared_ptr[CExpression]& sp):
+        Expression.init(self, sp)
+        self.scalar = <CFieldExpression*> sp.get()
+
+    def name(self):
+        return frombytes(self.scalar.name())
+
+
+cpdef enum CompareOperator:
+    Equal = <int8_t> CCompareOperator_EQUAL
+    NotEqual = <int8_t> CCompareOperator_NOT_EQUAL
+    Greater = <int8_t> CCompareOperator_GREATER
+    GreaterEqual = <int8_t> CCompareOperator_GREATER_EQUAL
+    Less = <int8_t> CCompareOperator_LESS
+    LessEqual = <int8_t> CCompareOperator_LESS_EQUAL
+
+
+cdef class ComparisonExpression(BinaryExpression):
+
+    cdef CComparisonExpression* comparison
+
+    def __init__(self, CompareOperator op,
+                 Expression left_operand not None,
+                 Expression right_operand not None):
+        cdef shared_ptr[CComparisonExpression] expression
+        expression.reset(
+            new CComparisonExpression(
+                <CCompareOperator>op,
+                left_operand.unwrap(),
+                right_operand.unwrap()
+            )
+        )
+        self.init(<shared_ptr[CExpression]> expression)
+
+    cdef void init(self, const shared_ptr[CExpression]& sp):
+        BinaryExpression.init(self, sp)
+        self.comparison = <CComparisonExpression*> sp.get()
+
+    def op(self):
+        return <CompareOperator> self.comparison.op()
+
+
+cdef class IsValidExpression(UnaryExpression):
+
+    def __init__(self, Expression operand not None):
+        cdef shared_ptr[CIsValidExpression] expression
+        expression = make_shared[CIsValidExpression](operand.unwrap())
+        self.init(<shared_ptr[CExpression]> expression)
+
+
+cdef class CastOptions:
```

Comment:
In the array.pxi code, we only expose safe=True/False for now, this should probably at point (not necessarily for this PR) be consolidated.

Reply:
Changed the CastExpression to receive the safe keyword, although I've kept and moved CastOptions to compute.pxi because we'll need it sooner or later. So we can add broader support for passing specific cast options in a single pass.